asm/f64/stubs_test.go

```diff
@@ -0,0 +1,689 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package f64
+
+import (
+	"math"
+	"testing"
+)
+
+var (
+	nan = math.NaN()
+	inf = math.Inf(1)
+)
+
+func guardVector(v []float64, g float64, g_ln int) (guarded []float64) {
+	guarded = make([]float64, len(v)+g_ln*2)
+	copy(guarded[g_ln:], v)
+	for i := 0; i < g_ln; i++ {
+		guarded[i] = g
+		guarded[len(guarded)-1-i] = g
+	}
+	return guarded
+}
+
+func isValidGuard(v []float64, g float64, g_ln int) bool {
+	for i := 0; i < g_ln; i++ {
+		if !same(v[i], g) || v[len(v)-1-i] != g {
+			return false
+		}
+	}
+	return true
+}
+
+func guardIncVector(vec []float64, guard_val float64, inc, guard_len int) (guarded []float64) {
+	s_ln := len(vec) * inc
+	guarded = make([]float64, s_ln+guard_len*2)
+	for i, j := 0, 0; i < len(guarded); i++ {
+		switch {
+		case i < guard_len, i > guard_len+s_ln:
```

Comment:
Conventionally we express this as `i < guard_len || guard_len+s_ln < i` for range checks (you can use `,` in place of `||`, but the order is important).